binderhub/tests/test_registry.py

```diff
@@ -0,0 +1,149 @@
+"""Tests for the registry"""
+import base64
+import json
+import os
+
+import pytest
+
+from tornado.web import Application, RequestHandler, HTTPError
+
+from binderhub.registry import DockerRegistry
+
+
+def test_registry_defaults(tmpdir):
+    registry = DockerRegistry(docker_config_path=str(tmpdir.join("doesntexist.json")))
+    assert registry.registry_host == "https://registry.hub.docker.com"
+    assert registry.auth_host == "https://index.docker.io/v1"
+    assert (
+        registry.token_url == "https://auth.docker.io/token?service=registry.docker.io"
+    )
+    assert registry.username == ""
+    assert registry.password == ""
+
+
+def test_registry_username_password(tmpdir):
+    config_json = tmpdir.join("dockerconfig.json")
+    with config_json.open("w") as f:
+        json.dump(
+            {
+                "auths": {
+                    "https://index.docker.io/v1": {
+                        "auth": base64.encodebytes(b"user:pass").decode("ascii")
+                    }
+                }
+            },
+            f,
+        )
+    registry = DockerRegistry(docker_config_path=str(config_json))
+    assert registry.username == "user"
+    assert registry.password == "pass"
+
+
+def test_registry_gcr_defaults(tmpdir):
+    config_json = tmpdir.join("dockerconfig.json")
+    with config_json.open("w") as f:
+        json.dump(
+            {
+                "auths": {
+                    "https://gcr.io": {
+                        "auth": base64.encodebytes(b"_json_key:{...}").decode("ascii")
+                    }
+                }
+            },
+            f,
+        )
+    registry = DockerRegistry(docker_config_path=str(config_json))
+    assert registry.registry_host == "https://gcr.io"
+    assert registry.auth_host == "https://gcr.io"
+    assert registry.token_url == "https://gcr.io/v2/token?service=gcr.io"
+    assert registry.username == "_json_key"
+    assert registry.password == "{...}"
```

Comment:
Do we need an additional test case where the user sets auth_host to docker and then sets registry host to gcr?

Reply:
I don't think so. That doesn't really make sense and won't work, but I'd rather let the underlying errors raise than try to preemptively detect such things, since these can be totally arbitrary URLs (private registries, amazon, etc.)

Reply:
I agree. I was looking at it from a slightly different perspective. The user gets confused about the registry host vs auth host and transposes them. Do we give reasonable messages for the user to not do it this way?